graphs/bidirectional_a_star.py

```diff
@@ -0,0 +1,223 @@
+        """
+        Retrace the path from parents to parents until start node
+        """
+        current_node = node
+        path = []
+        while current_node is not None:
+            path.append((current_node.pos_y, current_node.pos_x))
+            current_node = current_node.parent
+        path.reverse()
+        return path
+
+
+class BidirectionalAStar:
+    def __init__(self, start, goal):
+        self.fwd_astar = AStar(start, goal)
+        self.bwd_astar = AStar(goal, start)
+        self.reached = False
+        self.path = self.fwd_astar.path
+
+    def search(self):
+        while self.fwd_astar.open_nodes or self.bwd_astar.open_nodes:
+            self.fwd_astar.open_nodes.sort()
+            self.bwd_astar.open_nodes.sort()
+            current_fwd_node = self.fwd_astar.open_nodes.pop(0)
+            current_bwd_node = self.bwd_astar.open_nodes.pop(0)
+
+            if current_bwd_node.pos == current_fwd_node.pos:
+                self.reached = True
+                self.retrace_bidirectional_path(current_fwd_node, current_bwd_node)
+                break
+
+            self.fwd_astar.closed_nodes.append(current_fwd_node)
+            self.bwd_astar.closed_nodes.append(current_bwd_node)
+
+            self.fwd_astar.target = current_bwd_node
+            self.bwd_astar.target = current_fwd_node
+
+            successors = {
+                self.fwd_astar: self.fwd_astar.get_successors(current_fwd_node),
+                self.bwd_astar: self.bwd_astar.get_successors(current_bwd_node),
+            }
+
+            for astar in [self.fwd_astar, self.bwd_astar]:
+                for child_node in successors[astar]:
+                    if child_node in astar.closed_nodes:
+                        continue
+
+                    if child_node not in astar.open_nodes:
+                        astar.open_nodes.append(child_node)
+                    else:
+                        # retrieve the best current path
+                        better_node = astar.open_nodes.pop(
+                            astar.open_nodes.index(child_node)
+                        )
+
+                        if child_node.g_cost < better_node.g_cost:
+                            astar.open_nodes.append(child_node)
+                        else:
+                            astar.open_nodes.append(better_node)
+
+    def retrace_bidirectional_path(
+        self, fwd_node: Node, bwd_node: Node
+    ) -> List[Tuple[int]]:
+        fwd_path = self.fwd_astar.retrace_path(fwd_node)
+        bwd_path = self.bwd_astar.retrace_path(bwd_node)
+        fwd_path.reverse()
+        path = fwd_path + bwd_path
+        return path
+
+
+# all coordinates are given in format [y,x]
+init = (0, 0)
+goal = (len(grid) - 1, len(grid[0]) - 1)
+for elem in grid:
+    print(elem)
+
+start_time = time.time()
+a_star = AStar(init, goal)
+a_star.search()
+end_time = time.time() - start_time
+print("AStar execution time = ", end_time, " seconds")
```

Comment:
```suggestion
print(f"AStar execution time = {end_time:f} seconds")
```
Improve the ability to see very small floats.

Reply:
Oh, didn't know about that one, thanks !

Reply:
I do enjoy the "x.e-t" version though. Find it more readable than "0.00000x" to compare two values.